src/jit/importer.cpp

```diff
@@ -18269,4 +18271,325 @@ bool Compiler::IsMathIntrinsic(GenTreePtr tree)
+    if ((baseMethodAttribs & CORINFO_FLG_VIRTUAL) == 0)
+    {
+        assert(call->IsVirtualStub());
+        assert(opts.IsReadyToRun());
+        JITDUMP("--- base method not virtual, sorry\n");
+        return;
+    }
+
+    // If the objClass is sealed (final), then we may be able to devirtualize.
+    const DWORD objClassAttribs = info.compCompHnd->getClassAttribs(objClass);
+    const bool  objClassIsFinal = (objClassAttribs & CORINFO_FLG_FINAL) != 0;
+
+#if defined(DEBUG)
+    const char* const objClassNote  = objClassIsFinal ? " [final]" : "";
+    const char* const objClassName  = info.compCompHnd->getClassName(objClass);
+    const char* const baseClassName = info.compCompHnd->getClassName(baseClass);
+    const char*       baseMethodClassName;
+    const char* const baseMethodName = eeGetMethodName(baseMethod, &baseMethodClassName);
+
+    if (verbose)
+    {
+        printf("$$$ In %s: Maybe devirt?\n"
+               "    class for 'this' is %s%s (attrib %08x)\n"
+               "    class for method is %s\n"
+               "    base method is %s::%s\n",
+               info.compFullName, objClassName, objClassNote, objClassAttribs, baseClassName, baseMethodClassName,
+               baseMethodName);
+    }
+#endif // defined(DEBUG)
+
+    // Screen out interface calls.
+    // We might be able to devirtualize these some day.
+    const DWORD baseClassAttribs = info.compCompHnd->getClassAttribs(baseClass);
+    if ((baseClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        assert(call->IsVirtualStub());
+        JITDUMP("--- base class is interface, sorry\n");
+        return;
+    }
+
+    // Bail if obj class is an interface.
+    // See for instance System.ValueTuple`8::GetHashCode, where lcl 0 is System.IValueTupleInternal
+    //   IL_021d:  ldloc.0
+    //   IL_021e:  callvirt   instance int32 System.Object::GetHashCode()
+    if ((objClassAttribs & CORINFO_FLG_INTERFACE) != 0)
+    {
+        JITDUMP("--- obj class is interface, sorry\n");
+        return;
+    }
+
+    // Bail if obj class is a shared instance, in particular watch for System::__Canon.
+    // Perhaps overly broad...?
+    if ((objClassAttribs & CORINFO_FLG_SHAREDINST) != 0)
+    {
+        JITDUMP("--- obj class is shared instance class, sorry\n");
+        return;
+    }
+
+    // Fetch the method that would be called based on the declared type of 'this'
+    CORINFO_METHOD_HANDLE derivedMethod = info.compCompHnd->resolveVirtualMethod(baseMethod, objClass);
+
+    // If we failed to get a handle, we can't devirtualize.  This can
+    // happen when prejitting, if the devirtualization crosses
+    // servicing bubble boundaries.
+    if (derivedMethod == nullptr)
+    {
+        JITDUMP("--- no derived method, sorry\n");
+        return;
+    }
+
+    // Fetch attributes and do more vetting of the method. In general,
+    // if we can't inline, we won't devirtualize.
```

Comment:
This is a stale comment. I used to call `canInline` but have since moved the relevant checks to the VM side.